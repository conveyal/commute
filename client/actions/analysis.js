import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  const commuters = newAnalysis.commuters
  delete newAnalysis.commuters
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`),
    makeMockTrips(newAnalysis.id, organizationId, commuters)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organizationId) => [
  deleteLocally({ analysisId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]

const makeMockTrips = (analysisId, organizationId, commuters) => {
  const trips = []
  const modes = ['bike', 'car', 'transit', 'walk']
  const randRange = () => Math.random() * 0.4 + 0.8
  commuters.forEach((commuter) => {
    const curTrip = {
      commuterId: commuter.id
    }
    const tripMetricBase = Math.random()
    let bestValue = 9999999
    let bestMode
    modes.forEach((mode) => {
      // generate random trips
      const modePctMultiplier = randRange()
      const cost = tripMetricBase * 34.7 * modePctMultiplier * randRange()
      const distance = tripMetricBase * 69 * modePctMultiplier * randRange()
      const time = tripMetricBase * 7500 * modePctMultiplier * randRange()
      curTrip[mode] = {cost, distance, time}
      const tripValue = cost + time * 0.01
      if (tripValue < bestValue) bestMode = mode
    })
    curTrip.mostLikely = Object.assign({ mode: bestMode }, curTrip[bestMode])
    trips.push(curTrip)
  })
  return sendMockTrips({analysisId, organizationId, trips})
}
const sendMockTrips = createAction('receive mock calculated trips')
