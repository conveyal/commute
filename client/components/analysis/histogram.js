import React, {Component, PropTypes} from 'react'
import {Col, ControlLabel, FormControl, FormGroup, Grid, Panel, Row} from 'react-bootstrap'
import {
  Crosshair,
  DiscreteColorLegend,
  HorizontalGridLines,
  VerticalBarSeries,
  XAxis,
  YAxis
} from 'react-vis'

import ButtonLink from '../button-link'
import FlexiblePlot from '../flexible-plot'
import Icon from '../icon'
import analysisDefaults from '../../utils/analysisDefaults'
import {calcNumLessThan, getInitialSeries} from '../../utils/components'

const {metrics} = analysisDefaults
const metricsVals = Object.keys(metrics)

export default class Histogram extends Component {
  static propTypes = {
    // dispatch
    deleteAnalysis: PropTypes.func,

    // props
    analysis: PropTypes.object.isRequired,
    groupName: PropTypes.string.isRequired,
    site: PropTypes.object.isRequired
  }

  state = {
    crosshairValues: [],
    metric: 'time',
    series: getInitialSeries()
  }

  componentWillMount () {
    // calculate with selected metric
    this._calculateSeries({...this.state})
  }

  /**
   * Calculate series values given selected metric
   *
   * @param  {Object} newState The new state to work with
   */
  _calculateSeries (newState) {
    const arrayVals = this.props.analysis.tripVals
    const metricData = metrics[newState.metric]
    const multiplier = metricData.multiplier || 1

    // reset series data
    newState.series = getInitialSeries()

    // disallow transit distance metric
    if (newState.metric === 'distance') {
      newState.series = newState.series.filter((seriesMode) => seriesMode.mode !== 'transit')
    }

    newState.series = newState.series.map((seriesMode) => {
      seriesMode.data = []
      seriesMode.numInPreviousBins = 0
      return seriesMode
    })

    // iterate through all bins of histogram
    for (let i = 0; i < metricData.max; i += metricData.step) {
      const upper = i + metricData.step
      const curBinLabel = `${i} - ${upper}`

      // calculate number of trips by mode that fall inside current bin
      newState.series.forEach((seriesMode) => {
        // find total num to left
        const numLess = calcNumLessThan(arrayVals[seriesMode.mode][newState.metric],
          upper * multiplier)

        // add bin
        seriesMode.data.push({ x: curBinLabel, y: numLess - seriesMode.numInPreviousBins })

        // update for next iteration
        seriesMode.numInPreviousBins = numLess
      })
    }

    // add those off edge
    newState.series.forEach((seriesMode) => {
      // add bin with remnant
      seriesMode.data.push({
        x: `> ${metricData.max}`,
        y: arrayVals[seriesMode.mode][newState.metric].length - seriesMode.numInPreviousBins
      })
    })

    this.setState(newState)
  }

  /**
   * Format the title line of the crosshair.
   * @param {Array} values Array of values.
   * @returns {Object} The caption and the value of the title.
   * @private
   */
  _formatCrosshairTitle = (values) => {
    return {
      title: metrics[this.state.metric].unit,
      value: values[0].x
    }
  }

  /**
   * A callback to format the crosshair items.
   * @param {Object} values Array of values.
   * @returns {Array<Object>} Array of objects with titles and values.
   * @private
   */
  _formatCrosshairItems = (values) => {
    const {series} = this.state
    return values.map((v, i) => {
      return {
        title: series[i].title,
        value: v.y
      }
    })
  }

  /**
   * Handle the change of metric
   *
   * @param  {[type]} e [description]
   * @private
   */
  _handleMetricChange = (event) => {
    const newState = {...this.state}
    newState.metric = event.target.value
    this._calculateSeries(newState)
  }

  /**
   * Click handler for the legend.
   * @param {Object} item Clicked item of the legend.
   * @param {number} i Index of the legend.
   * @private
   */
  _legendClickHandler = (item, i) => {
    const {series} = this.state
    series[i].disabled = !series[i].disabled
    this.setState({series})
  }

  /**
   * Event handler for onMouseLeave.
   * @private
   */
  _mouseLeaveHandler = () => {
    this.setState({crosshairValues: []})
  }

  /**
   * Event handler for onNearestX.
   * @param {Object} value Selected value.
   * @param {number} index Index of the series.
   * @private
   */
  _nearestXHandler = (value, {index}) => {
    const {series} = this.state
    this.setState({
      crosshairValues: series.map(s => s.data[index])
    })
  }

  render () {
    const {_id: analysisId, name} = this.props.analysis
    const {groupName, site} = this.props
    const {crosshairValues, metric: selectedMetric, series} = this.state
    return (
      <Grid>
        <Row className='histogram-header'>
          <Col xs={12}>
            <h3>
              <span>{name}</span>
              <ButtonLink
                className='pull-right'
                to={`/analysis/${analysisId}`}
                >
                <Icon type='arrow-left' />
                <span>Back</span>
              </ButtonLink>
            </h3>
            <h3>Commute Metrics Histogram</h3>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Site:</strong>
              <span>{site.name}</span>
            </h4>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Group:</strong>
              <span>{groupName}</span>
            </h4>
          </Col>
        </Row>
        <Row>
          <Col xs={9}>
            <FlexiblePlot
              animation
              height={300}
              margin={{bottom: 70}}
              onMouseLeave={this._mouseLeaveHandler}
              xType='ordinal'
              >
              <Crosshair
                itemsFormat={this._formatCrosshairItems}
                titleFormat={this._formatCrosshairTitle}
                values={crosshairValues}
                />
              <HorizontalGridLines />
              {series
                .filter((s) => !s.disabled)
                .map((s, idx) =>
                  <VerticalBarSeries
                    color={s.color}
                    data={s.data}
                    key={`possibilities-series-${s.title}`}
                    onNearestX={this._nearestXHandler}
                    />
                )
              }
              <XAxis
                tickLabelAngle={-45}
                title={metrics[selectedMetric].unit}
                />
              <YAxis title='Number of Trips' />
            </FlexiblePlot>
          </Col>
          <Col xs={3}>
            <DiscreteColorLegend
              items={series}
              onItemClick={this._legendClickHandler}
              />
          </Col>
        </Row>
        <Row>
          <Col xs={12} md={8} className='histogram-settings'>
            <h4>Settings</h4>
            <Panel>
              <FormGroup controlId='histogram-metric-select'>
                <ControlLabel>Metric</ControlLabel>
                <FormControl
                  componentClass='select'
                  onChange={this._handleMetricChange}
                  value={selectedMetric}
                  >
                  {metricsVals.map((metric) => {
                    return <option value={metric}>{metrics[metric].label}</option>
                  })}
                </FormControl>
              </FormGroup>
            </Panel>
          </Col>
        </Row>
      </Grid>
    )
  }
}
