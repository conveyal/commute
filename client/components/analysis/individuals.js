import humanizeDuration from 'humanize-duration'
import {Browser} from 'leaflet'
import React, {Component, PropTypes} from 'react'
import {Button, Col, Grid, Row} from 'react-bootstrap'
import {BootstrapTable, TableHeaderColumn} from 'react-bootstrap-table'
import {Map as LeafletMap, TileLayer} from 'react-leaflet'
import {Link} from 'react-router'

import Icon from '../icon'
import {humanizeDistance} from '../../utils/components'
import {settings} from '../../utils/env'

export default class Individuals extends Component {
  static propTypes = {
    // dispatch
    deleteAnalysis: PropTypes.func,

    // props
    analysis: PropTypes.object.isRequired,
    commutersById: PropTypes.object.isRequired,
    groupName: PropTypes.string.isRequired,
    organizationId: PropTypes.string.isRequired,
    siteName: PropTypes.string.isRequired
  }

  constructor (props) {
    super(props)
    this.state = {
      position: settings.map.focus,
      zoom: 8
    }
  }

  _costRenderer = (cell, row) => {
    return `$${row.mostLikely.cost}`
  }

  _distanceRenderer = (cell, row) => {
    return humanizeDistance(row.mostLikely.distance)
  }

  _durationRenderer = (cell, row) => {
    return humanizeDuration(row.mostLikely.time * 60 * 1000)
  }

  _modeRenderer = (cell, row) => {
    return row.mostLikely.mode
  }

  _nameRenderer = (cell, row) => {
    return this.props.commutersById[row.commuterId].name
  }

  _onRowSelect = (row, isSelected) => {
    // TODO: show on map
    console.log(row)
    console.log('selected: ' + isSelected)
  }

  _ridematchesRenderer = (cell, row) => {
    return row.mostLikely.ridematches
  }

  render () {
    const {id, name} = this.props.analysis
    const {analysis, groupName, organizationId, siteName} = this.props
    const {position, zoom} = this.state
    return (
      <Grid>
        <Row className='individuals-header'>
          <Col xs={12}>
            <h3>
              <span>{name}</span>
              <Button className='pull-right'>
                <Link to={`/organizations/${organizationId}/analysis/${id}`}>
                  <Icon type='arrow-left' />
                  <span>Back</span>
                </Link>
              </Button>
            </h3>
            <h3>Individual Commutes</h3>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Site:</strong>
              <span>{siteName}</span>
            </h4>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Group:</strong>
              <span>{groupName}</span>
            </h4>
          </Col>
        </Row>
        <Row>
          <Col xs={8} style={{height: '400px'}}>
            <LeafletMap center={position} zoom={zoom}>
              <TileLayer
                url={Browser.retina &&
                  process.env.LEAFLET_RETINA_URL
                  ? process.env.LEAFLET_RETINA_URL
                  : process.env.LEAFLET_TILE_URL}
                attribution={process.env.LEAFLET_ATTRIBUTION}
                />
            </LeafletMap>
          </Col>
        </Row>
        <Row className='individuals-content'>
          <Col xs={12}>
            <h3>Individual Trips</h3>
            <BootstrapTable
              data={analysis.trips}
              selectRow={{
                mode: 'radio',
                clickToSelect: true,
                bgColor: 'rgb(238, 193, 213)',
                onSelect: this._onRowSelect
              }}
              >
              <TableHeaderColumn dataField='commuterId' isKey hidden />
              <TableHeaderColumn dataFormat={this._nameRenderer}>Name</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._modeRenderer}>Mode</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._durationRenderer}>Duration</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._distanceRenderer}>Distance</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._costRenderer}>Cost</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._ridematchesRenderer}>Ridematches</TableHeaderColumn>
            </BootstrapTable>
          </Col>
        </Row>
      </Grid>
    )
  }
}
