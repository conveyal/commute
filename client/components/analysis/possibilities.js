import humanizeDuration from 'humanize-duration'
import Slider from 'rc-slider'
import React, {Component, PropTypes} from 'react'
import {Button, Col, ControlLabel, FormControl, FormGroup, Grid, Panel, Row} from 'react-bootstrap'
import {Link} from 'react-router'
import {DiscreteColorLegend, HorizontalGridLines, VerticalBarSeries, YAxis} from 'react-vis'

import FlexiblePlot from '../flexible-plot'
import Icon from '../icon'
import {calcNumLessThan, getInitialSeries, humanizeDistance} from '../../utils/components'
import {settings} from '../../utils/env'

const METRICS = Object.keys(settings.metrics)

export default class Possibilities extends Component {
  static propTypes = {
    // dispatch
    deleteAnalysis: PropTypes.func,

    // props
    analysis: PropTypes.object.isRequired,
    groupName: PropTypes.string.isRequired,
    organizationId: PropTypes.string.isRequired,
    siteName: PropTypes.string.isRequired
  }

  constructor (props) {
    super(props)
    this.state = {
      cost: settings.metrics.cost.default,
      distance: settings.metrics.distance.default,
      series: getInitialSeries(),
      time: settings.metrics.time.default * 60,
      yAxisUnit: 'percent'
    }
  }

  componentWillMount () {
    // calculate with constraints
    this._calculateSeries({...this.state})
  }

  /**
   * Calculate series values given constraints for each mode
   *
   * @param  {Object} newState The new state to work with
   */
  _calculateSeries = (newState) => {
    const arrayVals = this.props.analysis.tripVals
    newState.series = newState.series.map((seriesMode) => {
      // populate all constraints
      const tripsByConstraint = []
      METRICS.forEach((metric) => {
        tripsByConstraint.push(calcNumLessThan(arrayVals[seriesMode.mode][metric], newState[metric]))
      })
      const constrainedNumber = Math.min.apply(this, tripsByConstraint)
      seriesMode.data = [{
        x: 1,
        y: (constrainedNumber /
          (newState.yAxisUnit === 'percent' ? this.props.analysis.trips.length * 0.01 : 1)
        ).toFixed(newState.yAxisUnit === 'percent' ? 1 : 0)
      }]
      return seriesMode
    })

    this.setState(newState)
  }

  _handleCostChange = (v) => {
    const newState = {...this.state}
    newState.cost = v
    this._calculateSeries(newState)
  }

  _handleDistanceChange = (v) => {
    const newState = {...this.state}
    newState.distance = v
    this._calculateSeries(newState)
  }

  _handleTimeChange = (v) => {
    const newState = {...this.state}
    newState.time = v * 60
    this._calculateSeries(newState)
  }

  _handleYAxisChange = (event) => {
    const newState = {...this.state}
    newState.yAxisUnit = event.target.value
    this._calculateSeries(newState)
  }

  /**
   * Click handler for the legend.
   * @param {Object} item Clicked item of the legend.
   * @param {number} i Index of the legend.
   * @private
   */
  _legendClickHandler = (item, i) => {
    const {series} = this.state
    series[i].disabled = !series[i].disabled
    this.setState({series})
  }

  render () {
    const {id, name} = this.props.analysis
    const {groupName, organizationId, siteName} = this.props
    const {series, yAxisUnit} = this.state
    const activeSeries = series.filter((s) => !s.disabled)
    return (
      <Grid>
        <Row className='possibilities-header'>
          <Col xs={12}>
            <h3>
              <span>{name}</span>
              <Button className='pull-right'>
                <Link to={`/organizations/${organizationId}/analysis/${id}`}>
                  <Icon type='arrow-left' />
                  <span>Back</span>
                </Link>
              </Button>
            </h3>
            <h3>Possibilities Analysis</h3>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Site:</strong>
              <span>{siteName}</span>
            </h4>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Group:</strong>
              <span>{groupName}</span>
            </h4>
          </Col>
        </Row>
        <Row>
          <Col xs={12} sm={9}>
            <FlexiblePlot
              animation
              height={300}
              >
              <HorizontalGridLines />
              {activeSeries
                .map((s, idx) =>
                  <VerticalBarSeries
                    color={s.color}
                    data={s.data}
                    key={`possibilities-series-${s.title}`}
                    />
                )
              }
              <YAxis title={yAxisUnit === 'percent' ? 'Percent' : 'Number of Commuters'} />
            </FlexiblePlot>
          </Col>
          <Col xs={12} sm={3} className='possibilities-legend-control'>
            <DiscreteColorLegend
              items={series}
              onItemClick={this._legendClickHandler}
              />
            <Panel>
              <FormGroup controlId='histogram-metric-select'>
                <ControlLabel>Y Axis</ControlLabel>
                <FormControl
                  componentClass='select'
                  onChange={this._handleYAxisChange}
                  value={yAxisUnit}
                  >
                  <option value='percent'>Percent</option>
                  <option value='number'>Number of Commuters</option>
                </FormControl>
              </FormGroup>
            </Panel>
          </Col>
          <Col xs={9}>
            <table className='table table-bordered'>
              <tbody>
                <tr>
                  {activeSeries
                    .map((s, idx) =>
                      <td key={idx}>
                        <span
                          className='rv-discrete-color-legend-item__color'
                          style={{background: s.color}}
                          />
                        <span
                          className='rv-discrete-color-legend-item__title'>
                          {`${s.title} - ${s.data[0].y}${yAxisUnit === 'percent' ? '%' : ''}`}
                        </span>
                      </td>
                    )
                  }
                </tr>
              </tbody>
            </table>
          </Col>
          <Col xs={12} className='possibilities-settings'>
            <h4>Settings</h4>
            <Panel>
              <p>Maximum Travel Time</p>
              <Slider
                defaultValue={settings.metrics.time.default}
                handle={
                  <CustomHandle
                    formatter={
                      // convert minutes to milliseconds
                      (v) => humanizeDuration(v * 60 * 1000)
                    }
                    />
                }
                max={settings.metrics.time.max}
                min={settings.metrics.time.min}
                onChange={this._handleTimeChange}
                />
            </Panel>
            <Panel>
              <p>Maximum Distance</p>
              <Slider
                defaultValue={settings.metrics.distance.default}
                handle={
                  <CustomHandle
                    formatter={humanizeDistance}
                    />}
                max={settings.metrics.distance.max}
                min={settings.metrics.distance.min}
                onChange={this._handleDistanceChange}
                />
            </Panel>
            <Panel>
              <p>Maximum Cost</p>
              <Slider
                defaultValue={settings.metrics.cost.default}
                handle={
                  <CustomHandle
                    formatter={(v) => `$${v}`}
                    />}
                max={settings.metrics.cost.max}
                min={settings.metrics.cost.min}
                onChange={this._handleCostChange}
                />
            </Panel>
          </Col>
        </Row>
      </Grid>
    )
  }
}

const handleStyle = {
  position: 'absolute',
  transform: 'translate(-50%, -50%)',
  cursor: 'pointer',
  padding: '2px',
  border: '2px solid #abe2fb',
  borderRadius: '3px',
  background: '#fff',
  fontSize: '14px',
  textAlign: 'center'
}

const CustomHandle = props => {
  const style = Object.assign({ left: `${props.offset}%` }, handleStyle)
  return (
    <div style={style}>{props.formatter(props.value)}</div>
  )
}

CustomHandle.propTypes = {
  formatter: PropTypes.func.isRequired,
  offset: PropTypes.number,
  value: PropTypes.any
}
