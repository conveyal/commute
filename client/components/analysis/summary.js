import humanizeDuration from 'humanize-duration'
import React, {Component, PropTypes} from 'react'
import {Button, Col, Grid, Row} from 'react-bootstrap'
import {Link} from 'react-router'

import Icon from '../icon'
import {humanizeDistance} from '../../utils/components'
import {messages} from '../../utils/env'
import {actUponConfirmation} from '../../utils/ui'

export default class Summary extends Component {
  static propTypes = {
    // dispatch
    deleteAnalysis: PropTypes.func,

    // props
    analysis: PropTypes.object.isRequired,
    groupName: PropTypes.string.isRequired,
    organizationId: PropTypes.string.isRequired,
    siteName: PropTypes.string.isRequired
  }

  handleDelete () {
    const {analysis, deleteAnalysis, organizationId} = this.props
    const doDelete = () => {
      deleteAnalysis(analysis.id, organizationId)
    }
    actUponConfirmation(messages.analysis.deleteConfirmation, doDelete)
  }

  render () {
    const {id, name, summary} = this.props.analysis
    const {groupName, organizationId, siteName} = this.props
    return (
      <Grid>
        <Row>
          <Col xs={12}>
            <h3>
              <span>{name}</span>
              <Button className='pull-right'>
                <Link to={`/organizations/${organizationId}`}><Icon type='arrow-left' />Back</Link>
              </Button>
            </h3>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Site:</strong>
              <span>{siteName}</span>
            </h4>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Group:</strong>
              <span>{groupName}</span>
            </h4>
          </Col>
          <Col xs={12}>
            <h4>
              <strong>Progress:</strong>
              <span>100%</span>
            </h4>
          </Col>
          <Col xs={12} md={4}>
            <h4>
              <strong>Potential Savings</strong>
            </h4>
          </Col>
          <Col xs={12} md={8}>
            <table className='table table-bordered'>
              <tbody>
                <tr>
                  <td>Total Costs Per Day</td>
                  <td>{`$${summary.savingsTotalPerDay}`}</td>
                </tr>
                <tr>
                  <td>Total Costs Per Year</td>
                  <td>{`$${summary.savingsTotalPerYear}`}</td>
                </tr>
                <tr>
                  <td>Cost Per Trip Per Day</td>
                  <td>{`$${summary.savingsPerTrip}`}</td>
                </tr>
                <tr>
                  <td>Cost Per Trip Per Year</td>
                  <td>{`$${summary.savingsPerTripYear}`}</td>
                </tr>
              </tbody>
            </table>
          </Col>
          <Col xs={12} md={4}>
            <h4>
              <strong>Mode Shift Stats</strong>
            </h4>
          </Col>
          <Col xs={12} md={8}>
            <table className='table table-bordered'>
              <tbody>
                <tr>
                  <td>Average Travel Time</td>
                  <td>{humanizeDuration(summary.avgTravelTime * 1000)}</td>
                </tr>
                <tr>
                  <td>Average Distance</td>
                  <td>{humanizeDistance(summary.avgDistance)}</td>
                </tr>
              </tbody>
            </table>
          </Col>
          <Col xs={12}>
            <p>
              <Link to={`/organizations/${organizationId}/analysis/${id}/histogram`}>
                Commute Metrics Histogram
              </Link>
            </p>
            <p>
              <Link to={`/organizations/${organizationId}/analysis/${id}/possibilities`}>
                Possibilities Analysis
              </Link>
            </p>
            <p>
              <Link to={`/organizations/${organizationId}/analysis/${id}/individuals`}>
                Individual Commuter Analysis
              </Link>
            </p>
          </Col>
        </Row>
      </Grid>
    )
  }
}
