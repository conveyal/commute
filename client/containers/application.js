import {connect} from 'react-redux'
import {refreshUser} from 'mastarm/react/auth0'

import Application from '../components/application'

function mapStateToProps (state) {
  return {}
}

function mapDispatchToProps (dispatch) {
  return {
    refreshUserToken: () => refreshUser(dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Application)
