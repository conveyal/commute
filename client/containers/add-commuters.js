import {connect} from 'react-redux'

import {appendToGroup, createGroup} from '../actions/group'
import AddCommuters from '../components/add-commuters'

function mapStateToProps (state, props) {
  const {organization} = state
  const {params} = props
  const currentOrganizationId = params.organizationId
  const output = {
    organizationId: currentOrganizationId
  }
  if (params && params.groupId) {
    const currentOrganization = organization.organizationsById[currentOrganizationId]
    const currentGroup = currentOrganization.groupsById[params.groupId]
    Object.assign(output, {
      appendMode: true,
      organizationId: currentOrganizationId,
      group: currentGroup
    })
  }
  return output
}

function mapDispatchToProps (dispatch, props) {
  return {
    create: (group, organizationId) =>
      dispatch(createGroup(group, organizationId)),
    append: (commuters, groupId, organizationId) =>
      dispatch(appendToGroup(commuters, groupId, organizationId))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddCommuters)
