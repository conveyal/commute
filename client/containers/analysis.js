import {connect} from 'react-redux'

import {deleteAnalysis} from '../actions/analysis'
import _histogram from '../components/analysis/histogram'
import _individuals from '../components/analysis/individuals'
import _possibilities from '../components/analysis/possibilities'
import _summary from '../components/analysis/summary'

function mapStateToProps (state, props) {
  const {organization} = state
  const {params} = props
  const currentOrganizationId = params.organizationId
  const currentOrganization = organization.organizationsById[currentOrganizationId]
  const curAnalysis = currentOrganization.analysesById[params.analysisId]
  const group = currentOrganization.groupsById[curAnalysis.groupId]
  return {
    analysis: curAnalysis,
    commutersById: group.commutersById,
    groupName: group.name,
    organizationId: currentOrganizationId,
    siteName: currentOrganization.sitesById[curAnalysis.siteId].name
  }
}

function mapDispatchToProps (dispatch) {
  return {
    deleteAnalysis: (id, organizationId) => dispatch(deleteAnalysis(id, organizationId))
  }
}

export const Histogram = connect(mapStateToProps)(_histogram)
export const Individuals = connect(mapStateToProps)(_individuals)
export const Possibilities = connect(mapStateToProps)(_possibilities)
export const Summary = connect(mapStateToProps, mapDispatchToProps)(_summary)
