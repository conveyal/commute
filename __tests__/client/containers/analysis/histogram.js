/* global describe, expect, it */

import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import React from 'react'
import {Provider} from 'react-redux'

import {makeMockStore, mockStores} from '../../../test-utils/mock-data.js'

import {Histogram} from '../../../../client/containers/analysis'

const mockStore = makeMockStore(mockStores.complexOrganization)

describe('Container > Analysis > Histogram', () => {
  it('Histogram View loads (base case)', () => {
    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <Histogram
          params={{organizationId: '2', analysisId: '1'}}
          />
      </Provider>
    )
    expect(mountToJson(tree.find('.histogram-header'))).toMatchSnapshot()
    expect(mountToJson(tree.find('.histogram-settings'))).toMatchSnapshot()
  })

  /* it('Toggle mode checkbox', () => {
    // Given a logged-in user is viewing the Commute Time by Mode View
    // When the user toggles the checkbox to display a particular mode in the histogram
    // Then the histogram should redraw itself with the particular mode display state
  }) */
})
