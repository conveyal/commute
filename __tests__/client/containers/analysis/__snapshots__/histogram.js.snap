exports[`Container > Analysis > Histogram Histogram View loads (base case) 1`] = `
<div
  className="histogram-header row">
  <Col
    bsClass="col"
    componentClass="div"
    xs={12}>
    <div
      className="col-xs-12">
      <h3>
        <span>
          An Analysis
        </span>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="pull-right"
          disabled={false}>
          <button
            className="pull-right btn btn-default"
            disabled={false}
            type="button">
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/organizations/2/analysis/1">
              <a
                onClick={[Function]}
                style={Object {}}>
                <Icon
                  type="arrow-left">
                  <i
                    className="fa fa-arrow-left fa-fw " />
                </Icon>
                <span>
                  Back
                </span>
              </a>
            </Link>
          </button>
        </Button>
      </h3>
      <h3>
        Commute Metrics Histogram
      </h3>
    </div>
  </Col>
  <Col
    bsClass="col"
    componentClass="div"
    md={6}
    xs={12}>
    <div
      className="col-md-6 col-xs-12">
      <h4>
        <strong>
          Site:
        </strong>
        <span>
          Acme Corp
        </span>
      </h4>
    </div>
  </Col>
  <Col
    bsClass="col"
    componentClass="div"
    md={6}
    xs={12}>
    <div
      className="col-md-6 col-xs-12">
      <h4>
        <strong>
          Group:
        </strong>
        <span>
          Mock Group
        </span>
      </h4>
    </div>
  </Col>
</div>
`;

exports[`Container > Analysis > Histogram Histogram View loads (base case) 2`] = `
<div
  className="histogram-settings col-md-8 col-xs-12">
  <h4>
    Settings
  </h4>
  <Panel
    bsClass="panel"
    bsStyle="default"
    defaultExpanded={false}>
    <div
      className="panel panel-default"
      id={undefined}>
      <div
        className="panel-body">
        <FormGroup
          bsClass="form-group"
          controlId="histogram-metric-select">
          <div
            className="form-group">
            <ControlLabel
              bsClass="control-label"
              srOnly={false}>
              <label
                className="control-label"
                htmlFor="histogram-metric-select">
                Metric
              </label>
            </ControlLabel>
            <FormControl
              bsClass="form-control"
              componentClass="select"
              onChange={[Function]}
              value="time">
              <select
                className="form-control"
                id="histogram-metric-select"
                onChange={[Function]}
                type={undefined}
                value="time">
                <option
                  value="cost">
                  cost
                </option>
                <option
                  value="distance">
                  distance
                </option>
                <option
                  value="time">
                  time
                </option>
              </select>
            </FormControl>
          </div>
        </FormGroup>
      </div>
    </div>
  </Panel>
</div>
`;
